### PR TITLE
WIP - Multi-CassKop implementation

### PR DESCRIPTION
**Pull request by allamand**
This PR goal is to add new sub-project multi-Casskop in the directory with the same name.

This indroduce new Operator which manage MultiCasskop CRD, and connects to several K8s clusters to create CassandraCluster crd instance in each. From there Casskop will create Cassandra clusters based on the CassandraCluster object created by multi-casskop.

### CODE REVIEW

**Comment by cscetbon** [comment]
@allamand The other branch fixing the cleanup has been merged. You can skip it on that branch

File: multi-casskop/Readme.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+time="2019-11-28T15:50:19Z" level=info msg="Cluster is not Ready, we requeue [phase=Initializing / action=Initializing / status=Ongoing]" cluster=cassandra-e2e kubernetes=k8s-cluster2 namespace=cassandra-e2e
+time="2019-11-28T15:50:49Z" level=info msg="Cluster is not Ready, we requeue [phase=Initializing / action=Initializing / status=Ongoing]" cluster=cassandra-e2e kubernetes=k8s-cluster2 namespace=cassandra-e2e
+time="2019-11-28T15:51:19Z" level=info msg="Cluster is not Ready, we requeue [phase=Initializing / action=Initializing / status=Ongoing]" cluster=cassandra-e2e kubernetes=k8s-cluster2 namespace=cassandra-e2e
+time="2019-11-28T15:51:49Z" level=info msg="Cluster is not Ready, we requeue [phase=Initializing / action=Initializing / status=Ongoing]" cluster=cassandra-e2e kubernetes=k8s-cluster2 namespace=cassandra-e2e
+```
+
+This is the sequence of operations:
+- MultiCassKop first creates the CassandraCluster in k8s-cluster1. 
+- Then local CassKop started to creates the associated Cassandra Cluster.
+  - When CassKop ended created it's cluster, it updates the CassandraCluster status with the phase=Running meaning that
```
Comment (cscetbon) [inline multi-casskop/Readme.md]: When CasskOp has created its Cassandra Cluster, it updates CassandraCluster object 's status with phase=Running meaning ...

File: multi-casskop/TODO.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+
+
+## Features
+
+- [ ] **Feature01**: the CassKop special feature using the _unlockNextOperation which allows to recover when CassKop fails,
```
Comment (cscetbon) [inline multi-casskop/TODO.md]: CassKop's special feature ...

File: multi-casskop/cmd/manager/main.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+const (
+	logLevelEnvVar = "LOG_LEVEL"
+)
+
+func getLogLevel() logrus.Level {
+	logLevel, found := os.LookupEnv(logLevelEnvVar)
+	if !found {
+		return logrus.InfoLevel
+	}
+	switch strings.ToUpper(logLevel) {
```
Comment (cscetbon) [inline multi-casskop/cmd/manager/main.go]: Good that you take this fix ;)

File: multi-casskop/deploy/crds/multicluster_v1alpha1_cassandramulticluster_cr.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      name: cassandra-demo
+      namespace: cassandra-demo
+      labels:
+        cluster: casskop
+    spec:
+      resources:         
+        requests:
+          cpu: '0.3'
+          memory: 2Gi
+        limits:
```
Comment (cscetbon) [inline multi-casskop/deploy/crds/multicluster_v1alpha1_cassandramulticluster_cr.yaml]: You can simplify the yaml like I did in the past
```
requests: &requests
  cpu: '0.3'
  memory: 2Gi
limits: *requests
```

File: multi-casskop/deploy/operator.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  template:
+    metadata:
+      labels:
+        name: multi-casskop
+    spec:
+      serviceAccountName: multi-casskop
+      containers:
+        - name: multi-casskop
+          # Replace this with the built image name
+          image: REPLACE_IMAGE
```
Comment (cscetbon) [inline multi-casskop/deploy/operator.yaml]: Let's just put the comment in here and remove it on line 18
`image: "Replace this with the built image name"`

File: multi-casskop/documentation/version-alpha.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+## What is Multi-CassKop
+
+Multi-CassKop is a new Kubernetes operator that sits above CassKop operator.
+CassKop allows to create a Cassandra cluster in a Kubernetes cluster, using the custom ressource **CassandraCluster**.
+
+Multi-CassKop goal is to bring the ability to deploy a Cassandra cluster within different regions, each of them running
+an independant Kubernetes cluster.
+
+Multi-CassKop will be talking to several Kubernetes clusters APIs, and is managing a new custom ressource named
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: is managing -> manage

File: multi-casskop/documentation/version-alpha.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+**MultiCasskop**.
+
+The only goal of Multi-CassKop is to create **CassandraCluster** ressources in each of k8s clusters, where there is a
+locally instance of CassKop that will deploy Cassandra cluster according to the required definition.
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: locally -> local

File: multi-casskop/documentation/version-alpha.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+Multi-Casskop insure that the Cassandra nodes deployed by each local CassKop will be part of the same Cassandra ring by
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: insure -> ensures

File: multi-casskop/documentation/version-alpha.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+This operator uses: 
+- Operator sdk to setup the new CRD `MultiCasskop`.
+- Admiralty's [multicluster-controller](https://github.com/admiraltyio/multicluster-controller) to allow talking to
+  several kubernetes api servers.
+  
+
+## Launching MultiCassKop operator
+
+MultiCassKop take into parameters a list of kubernetes contexts names as defined in the KUBECONFIG file.
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: take -> takes
contexts -> context

File: multi-casskop/documentation/version-alpha.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+example: 
+```
+./multiCassKop dex-sallamand-kaas-prod-priv-sph dex-sallamand-kaas-prod-priv-bgl
+```
+
+>**Note:** MultiCassKop will ONLY uses the first kubernetes passed as parameter to MultiCasskop objects
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: uses -> use
You can remove `as parameter`

File: multi-casskop/documentation/version-alpha.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+# How MultiCassKop works
+
+MultiCassKop starts by iterrating on every contexts passed in parameters then it register the controller.
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: iterating -> iterating
contexts -> context
same comment as before too
register -> registers

File: multi-casskop/documentation/version-alpha.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+```
+./multiCassKop dex-sallamand-kaas-prod-priv-sph dex-sallamand-kaas-prod-priv-bgl
+```
+
+>**Note:** MultiCassKop will ONLY uses the first kubernetes passed as parameter to MultiCasskop objects
+
+# How MultiCassKop works
+
+MultiCassKop starts by iterrating on every contexts passed in parameters then it register the controller. 
+The controller needs to be able to interract with MultiCasskop and CassandraCluster CRD objetcs.
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: interract -> interact

File: multi-casskop/documentation/version-alpha.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+In addition the controller needs to watch for MultiCasskop as it will need to react on any changes occurs on
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: occurs -> that occurred

File: multi-casskop/documentation/version-alpha.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+thoses objects for the given namespace.
+
+
+## MultiCasskop definition
+
+Multi-CassKop introduce a new custom ressource and will have the charge to create CassandraCluster ressoruces in each
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: introduce -> introduces
resources

File: multi-casskop/documentation/version-alpha.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+k8s cluster.
+
+The Spec field of MultiCasskop has a `base` parameter which contain a valid CassandraCluster object.
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: contains

File: multi-casskop/documentation/version-alpha.md [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+It also have an `override` section that will allow specify part of the base CassandraCluster definition that will be
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: have -> has
that will be override -> to be overridden

File: multi-casskop/documentation/version-alpha.md [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+override depending on the target cluster.
+
+Multi-CassKop make uses of kubectl context to target different clusters, so that the key of each override section must
```
Comment (cscetbon) [inline multi-casskop/documentation/version-alpha.md]: make uses -> makes use
context -> contexts

File: Changelog.md [hunk 18]
```diff
@@ -2,20 +2,30 @@
...
+## 0.5.0
+
+Introduce Multi-Casskop, the Operator to manage a single Cassandra Cluster above multiple Kubernetes clusters.
+
+- PR [#145](https://github.com/Orange-OpenSource/cassandra-k8s-operator/pull/145) - Fix [Issue #142](https://github.com/Orange-OpenSource/cassandra-k8s-operator/issues/142) PodStatus which rarely fails in
+  unit tests
+- PR [#146](https://github.com/Orange-OpenSource/cassandra-k8s-operator/pull/146) - Fix [Issue #143](https://github.com/Orange-OpenSource/cassandra-k8s-operator/issues/143) External update of SeedList was not possible
+
+- PR [#147](https://github.com/Orange-OpenSource/cassandra-k8s-operator/pull/147) - Introduce Multi-CassKop operator
+
```
[the remark this answers is not in the document]
Comment (allamand) [reply Changelog.md]: added

File: documentation/development.md [hunk 19]
```diff
@@ -330,6 +330,20 @@ and let's the magic happened
...
+- [ ] Update Changelog.md with informations for the new release
+- [ ] update version/version.go with the new release version
+- [ ] update multi-casskop/version/version.go with the new release version
+- [ ] update helm/cassandra-operator/Chart.yaml and values.yaml
+- [ ] update multi-casskop/helm/multi-casskop/Chart.yaml and values.yaml
+- [ ] generate casskop helm with `make helm-package`
+- [ ] add to git docs/helm, commit & push
+- [ ] once the PR is merged to master, create the release with content of changelog for this version
+    - https://github.com/Orange-OpenSource/cassandra-k8s-operator/releases
+
```
[the remark this answers is not in the document]
Comment (allamand) [reply documentation/development.md]: there are some bots that can help on this

File: Makefile [hunk 20]
```diff
@@ -234,12 +235,14 @@ debug-pod-logs:
 	kubectl logs -f `kubectl get pod -l app=cassandra-k8s-operator -o jsonpath="{.items[0].metadata.name}"`
 
 define debug_telepresence
-	export TELEPRESENCE_REGISTRY=$(TELEPRESENCE_REGISTRY)
-	echo "execute : cat cassandra-operator.env"
-	sudo mkdir -p /var/run/secrets/kubernetes.io
-	sudo ln -s /tmp/known/var/run/secrets/kubernetes.io/serviceaccount /var/run/secrets/kubernetes.io/
-	tdep=`kubectl get deployment -l app=cassandra-operator -o jsonpath='{.items[0].metadata.name}'`
-	telepresence --swap-deployment $$tdep --mount=/tmp/known --env-file cassandra-operator.env $1 $2
+	export TELEPRESENCE_REGISTRY=$(TELEPRESENCE_REGISTRY) ; \
```
[the remark this answers is not in the document]
Comment (allamand) [reply Makefile]: it's to debug the operator in cluster, and we need the env if we need to change the telepresence image

File: multi-casskop/Readme.md [hunk 21]
```diff
@@ -0,0 +1,247 @@
+
+# Multi-CassKop Project
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
```
[the remark this answers is not in the document]
Comment (allamand) [reply multi-casskop/Readme.md]: in emacs ^^

File: multi-casskop/documentation/version-alpha.md (hunk 13 above)
Comment (allamand) [reply multi-casskop/documentation/version-alpha.md]: oh forgot about this file, i'll read it again to see what need to be changed

File: multi-casskop/cmd/manager/main.go (hunk 3 above)
Comment (allamand) [reply multi-casskop/cmd/manager/main.go]: ^^

File: multi-casskop/deploy/crds/multicluster_v1alpha1_cassandramulticluster_cr.yaml (hunk 4 above)
Comment (allamand) [reply multi-casskop/deploy/crds/multicluster_v1alpha1_cassandramulticluster_cr.yaml]: this file is generated

File: multi-casskop/deploy/operator.yaml (hunk 5 above)
Comment (allamand) [reply multi-casskop/deploy/operator.yaml]: wonder if I don't have a script that uses the REPLACE_IMAGE, I prefer keep it for now

**Pull request merged**